File: packages/core-py/src/ebb_ai/routing.py
"""Cross-provider LLM routing — scored-candidate selection (Python mirror of
``routing.ts``, ROADMAP item 1).

Given a caller-allowed set of ``provider:model`` candidates and the grid
intensity at the already-chosen dispatch window, pick the best candidate on a
weighted blend of carbon, cost and latency. The window is chosen once by
``select_window`` BEFORE this runs — routing never shops for a per-candidate
window.

Honesty constraints (mirrored in the reasoning string):

1. No silent model swaps — routing only picks among the caller's candidates;
   there are no built-in equivalence tiers.
2. Grid honesty — hosted candidates are all scored against the SAME
   caller's-grid intensity (the receipt's documented assumption); carbon is
   differentiated via per-model energy coefficients, not pretend per-provider
   grids. Ollama is genuinely local.
3. Explainable pick — the full scored list + normalized weights are returned
   (and recorded on the receipt); the pick is reproducible given the rng seed.

See ``packages/core-ts/src/routing.ts`` for the full rationale.
"""

from __future__ import annotations

import math
import random
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any

from . import _data
from .energy import estimate_energy_kwh, normalize_model_name
from .types import Provider

#: Default weights when the caller does not supply ``route_weights``.
DEFAULT_ROUTE_WEIGHTS: dict[str, float] = {"carbon": 0.6, "cost": 0.3, "latency": 0.1}

#: Ties within this score epsilon are broken with the rng (reproducible given a seed).
SCORE_TIE_EPSILON = 1e-9

#: Disclosure prefix on a preview's reasoning line.
ROUTING_PREVIEW_DISCLOSURE = (
    "PREVIEW — the binding pick is decided at schedule time and may differ "
    "if the forecast shifts before commit"
)

_KNOWN_PROVIDERS = frozenset({"anthropic", "openai", "gemini", "ollama"})
_BATCH_CAPABLE_PROVIDERS = frozenset({"anthropic", "openai"})

_TYPICAL_INPUT_TOKENS: int = _data.TYPICAL_INPUT_TOKENS
_TYPICAL_OUTPUT_TOKENS: int = _data.TYPICAL_OUTPUT_TOKENS
#: Per-model public list prices (USD per Mtok), keyed by canonical model id.
MODEL_PRICES: dict[str, dict[str, Any]] = _data.MODEL_PRICES


class InvalidCandidateError(ValueError):
    """A ``provider:model`` candidate string is malformed."""


class MissingPriceError(ValueError):
    """One or more candidate models are absent from the SSOT price table.

    Loud by design — routing never guesses a price.
    """

    def __init__(self, missing: list[str]) -> None:
        self.missing = missing
        super().__init__(
            "No price in data/prices.json for routing candidate model(s): "
            + ", ".join(missing)
            + ". Add each to prices.json (with asOf + source) or drop it from the "
            "candidate list — routing never guesses a price."
        )


class InvalidRouteWeightsError(ValueError):
    """The caller-supplied weights cannot be normalized."""


@dataclass(frozen=True, slots=True)
class RoutingCandidate:
    """A parsed ``provider:model`` candidate."""

    provider: Provider
    model: str


@dataclass(slots=True)
class ScoredCandidate:
    """One fully-scored candidate, as recorded on the receipt's routing block."""

    provider: Provider
    model: str
    est_carbon_g: float
    est_cost_usd: float
    latency_class: float
    score: float

    def to_camel_dict(self) -> dict[str, Any]:
        return {
            "provider": self.provider,
            "model": self.model,
            "estCarbonG": self.est_carbon_g,
            "estCostUsd": self.est_cost_usd,
            "latencyClass": self.latency_class,
            "score": self.score,
        }

    def to_snake_dict(self) -> dict[str, Any]:
        return {
            "provider": self.provider,
            "model": self.model,
            "est_carbon_g": self.est_carbon_g,
            "est_cost_usd": self.est_cost_usd,
            "latency_class": self.latency_class,
            "score": self.score,
        }


@dataclass(slots=True)
class RoutingDecision:
    """The routing decision — persisted at schedule time, embedded in the
    signed receipt at completion. ``chosen`` / ``fallback_from`` are compact
    ``provider:model`` ids."""

    weights: dict[str, float]
    considered: list[ScoredCandidate]
    chosen: str
    reasoning: str
    fallback_from: str | None = None
    #: True for a NON-BINDING preview (recommend_window / dry_run). Never set
    #: on a committed / receipt-embedded decision.
    preview: bool = False

    def to_camel_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {
            "weights": {
                "carbon": self.weights["carbon"],
                "cost": self.weights["cost"],
                "latency": self.weights["latency"],
            },
            "considered": [c.to_camel_dict() for c in self.considered],
            "chosen": self.chosen,
            "reasoning": self.reasoning,
        }
        if self.fallback_from is not None:
            d["fallbackFrom"] = self.fallback_from
        if self.preview:
            d["preview"] = True
        return d

    def to_snake_dict(self) -> dict[str, Any]:
        """snake_case rendering for the planning-surface JSON payloads
        (recommend_window / dry_run), matching the TS ``routingBlockPayload``."""
        d: dict[str, Any] = {}
        if self.preview:
            d["preview"] = True
        d["chosen"] = self.chosen
        if self.fallback_from is not None:
            d["fallback_from"] = self.fallback_from
        d["weights"] = {
            "carbon": self.weights["carbon"],
            "cost": self.weights["cost"],
            "latency": self.weights["latency"],
        }
        d["considered"] = [c.to_snake_dict() for c in self.considered]
        d["reasoning"] = self.reasoning
        return d


def parse_candidate(spec: str) -> RoutingCandidate:
    """Parse a single ``"provider:model"`` string."""
    if not isinstance(spec, str):
        raise InvalidCandidateError(f"Invalid routing candidate {spec!r}: expected a string")
    trimmed = spec.strip()
    idx = trimmed.find(":")
    if idx <= 0:
        raise InvalidCandidateError(f'Invalid routing candidate {spec!r}: expected "provider:model"')
    provider = trimmed[:idx].strip().lower()
    model = trimmed[idx + 1 :].strip()
    if provider not in _KNOWN_PROVIDERS:
        raise InvalidCandidateError(
            f"Invalid routing candidate {spec!r}: unknown provider {provider!r} "
            "(expected one of anthropic, openai, gemini, ollama)"
        )
    if not model:
        raise InvalidCandidateError(f"Invalid routing candidate {spec!r}: model part is empty")
    return RoutingCandidate(provider=provider, model=model)  # type: ignore[arg-type]


def parse_candidates(specs: list[str]) -> list[RoutingCandidate]:
    return [parse_candidate(s) for s in specs]


def candidate_id(provider: str, model: str) -> str:
    return f"{provider}:{model}"


def normalize_route_weights(w: dict[str, float] | None = None) -> dict[str, float]:
    """Normalize caller weights to a non-negative vector summing to 1. ``None``
    ⇒ :data:`DEFAULT_ROUTE_WEIGHTS`. Missing keys default to 0; any negative
    value or an all-zero vector is rejected."""
    if w is None:
        return dict(DEFAULT_ROUTE_WEIGHTS)
    carbon = w.get("carbon", 0) or 0
    cost = w.get("cost", 0) or 0
    latency = w.get("latency", 0) or 0
    for name, v in (("carbon", carbon), ("cost", cost), ("latency", latency)):
        if not isinstance(v, (int, float)) or isinstance(v, bool) or math.isnan(v):
            raise InvalidRouteWeightsError(f"{name} must be a number, got {v!r}")
        if v < 0:
            raise InvalidRouteWeightsError(f"{name} must be non-negative, got {v}")
    total = carbon + cost + latency
    if total <= 0:
        raise InvalidRouteWeightsError("at least one weight must be positive")
    return {"carbon": carbon / total, "cost": cost / total, "latency": latency / total}


def price_for_model(model: str) -> dict[str, Any] | None:
    """Look up the SSOT price for a model id, trying exact then normalized."""
    exact = MODEL_PRICES.get(model.strip().lower())
    if exact is not None:
        return exact
    return MODEL_PRICES.get(normalize_model_name(model))


def _will_run_batch(
    candidate: RoutingCandidate,
    price: dict[str, Any],
    batch_eligible: bool,
    batch_capable: Callable[[str], bool],
) -> bool:
    return (
        batch_eligible
        and candidate.provider in _BATCH_CAPABLE_PROVIDERS
        and price.get("batch_discount") is not None
        and batch_capable(candidate.provider)
    )


def _latency_class_for(candidate: RoutingCandidate, runs_batch: bool) -> float:
    if candidate.provider == "ollama":
        return 0.0
    return 1.0 if runs_batch else 0.5


def _cost_usd_for(
    price: dict[str, Any], input_tokens: int, output_tokens: int, runs_batch: bool
) -> float:
    base = (input_tokens / 1_000_000) * price["in_usd_per_mtok"] + (
        output_tokens / 1_000_000
    ) * price["out_usd_per_mtok"]
    discount = price.get("batch_discount")
    mult = discount if (runs_batch and discount is not None) else 1
    return base * mult


def _round(v: float, places: int) -> float:
    factor = 10**places
    return math.floor(v * factor + 0.5) / factor


def _fmt_score(x: float) -> str:
    s = f"{x:.2f}"
    return s[1:] if s.startswith("0.") else s


def score_candidates(
    candidates: list[RoutingCandidate],
    intensity_g_co2_per_kwh: float,
    *,
    weights: dict[str, float] | None = None,
    batch_eligible: bool = False,
    batch_capable: Callable[[str], bool] | None = None,
    input_tokens: int | None = None,
    output_tokens: int | None = None,
    rng: Callable[[], float] | None = None,
) -> RoutingDecision:
    """Score a candidate set and pick the lowest-scoring one.

    Raises :class:`MissingPriceError` if any candidate model is absent from the
    price table (loud, never guessed).
    """
    if not candidates:
        raise ValueError("score_candidates: candidates must be non-empty")
    if batch_capable is None:
        batch_capable = lambda p: p in _BATCH_CAPABLE_PROVIDERS  # noqa: E731
    if input_tokens is None:
        input_tokens = _TYPICAL_INPUT_TOKENS
    if output_tokens is None:
        output_tokens = _TYPICAL_OUTPUT_TOKENS
    if rng is None:
        rng = random.random
    norm_weights = normalize_route_weights(weights)

    # Resolve every price FIRST so a missing one rejects the whole set loudly.
    prices: list[dict[str, Any]] = []
    missing: list[str] = []
    for c in candidates:
        p = price_for_model(c.model)
        if p is None:
            missing.append(candidate_id(c.provider, c.model))
        else:
            prices.append(p)
    if missing:
        raise MissingPriceError(missing)

    raw = []
    for c, price in zip(candidates, prices, strict=True):
        runs_batch = _will_run_batch(c, price, batch_eligible, batch_capable)
        carbon_g = (
            estimate_energy_kwh(
                model=c.model, input_tokens=input_tokens, output_tokens=output_tokens
            )
            * intensity_g_co2_per_kwh
        )
        cost_usd = _cost_usd_for(price, input_tokens, output_tokens, runs_batch)
        latency_class = _latency_class_for(c, runs_batch)
        raw.append((c, carbon_g, cost_usd, latency_class))

    def _norm(values: list[float]) -> list[float]:
        lo = min(values)
        hi = max(values)
        if hi - lo <= 0:
            return [0.0 for _ in values]
        return [(v - lo) / (hi - lo) for v in values]

    n_carbon = _norm([r[1] for r in raw])
    n_cost = _norm([r[2] for r in raw])
    n_latency = _norm([r[3] for r in raw])

    considered: list[ScoredCandidate] = []
    for i, (c, carbon_g, cost_usd, latency_class) in enumerate(raw):
        score = (
            norm_weights["carbon"] * n_carbon[i]
            + norm_weights["cost"] * n_cost[i]
            + norm_weights["latency"] * n_latency[i]
        )
        considered.append(
            ScoredCandidate(
                provider=c.provider,
                model=c.model,
                est_carbon_g=_round(carbon_g, 4),
                est_cost_usd=_round(cost_usd, 6),
                latency_class=latency_class,
                score=_round(score, 6),
            )
        )

    min_score = min(c.score for c in considered)
    tied = [c for c in considered if c.score - min_score <= SCORE_TIE_EPSILON]
    chosen = tied[math.floor(rng() * len(tied))] if len(tied) > 1 else tied[0]

    return RoutingDecision(
        weights=norm_weights,
        considered=considered,
        chosen=candidate_id(chosen.provider, chosen.model),
        reasoning=_build_reasoning(considered, chosen, norm_weights),
    )


def preview_routing(
    candidates: list[str] | None,
    intensity_g_co2_per_kwh: float,
    *,
    weights: dict[str, float] | None = None,
    batch_eligible: bool = False,
    rng: Callable[[], float] | None = None,
) -> RoutingDecision | None:
    """Non-binding routing preview for the planning surfaces
    (``recommend_window``, ``schedule_task`` dry_run). Runs the SAME scoring
    as :func:`score_candidates` at the previewed window's intensity, then
    marks the result ``preview=True`` and prefixes the reasoning with
    :data:`ROUTING_PREVIEW_DISCLOSURE`. Returns ``None`` when fewer than two
    candidates were supplied (routing is a no-op — so the params are never
    inert: they take effect exactly when >= 2 candidates are present). Raises
    :class:`MissingPriceError` / :class:`InvalidCandidateError` loudly, same
    as the committing path.
    """
    if not isinstance(candidates, list) or len(candidates) < 2:
        return None
    decision = score_candidates(
        parse_candidates(candidates),
        intensity_g_co2_per_kwh,
        weights=weights,
        batch_eligible=batch_eligible,
        rng=rng,
    )
    decision.preview = True
    decision.reasoning = f"{ROUTING_PREVIEW_DISCLOSURE}: {decision.reasoning}"
    return decision


def _is_min_by(all_c: list[ScoredCandidate], target: ScoredCandidate, key) -> bool:
    lo = min(key(c) for c in all_c)
    return key(target) <= lo + SCORE_TIE_EPSILON


def _build_reasoning(
    considered: list[ScoredCandidate],
    chosen: ScoredCandidate,
    weights: dict[str, float],
) -> str:
    others = [c for c in considered if c is not chosen]
    vs = (
        f" (score {_fmt_score(chosen.score)} vs "
        + "/".join(_fmt_score(o.score) for o in others)
        + ")"
        if others
        else ""
    )
    clauses: list[str] = []
    if weights["carbon"] > 0 and _is_min_by(considered, chosen, lambda c: c.est_carbon_g):
        clauses.append("lowest carbon at window")
    if weights["cost"] > 0 and _is_min_by(considered, chosen, lambda c: c.est_cost_usd):
        clauses.append("no per-token cost (local)" if chosen.est_cost_usd == 0 else "lowest cost")
    if weights["latency"] > 0 and _is_min_by(considered, chosen, lambda c: c.latency_class):
        clauses.append(
            "local (no network latency tier)"
            if chosen.latency_class == 0
            else "fastest latency tier"
        )
    if not clauses:
        clauses.append("best weighted blend of carbon/cost/latency")
    grid_honesty = (
        "runs locally on your own grid"
        if chosen.provider == "ollama"
        else "hosted-grid assumption applies (all hosted candidates scored at the same caller's-grid intensity)"
    )
    return (
        f"routed to {candidate_id(chosen.provider, chosen.model)}{vs}: "
        + "; ".join(clauses)
        + f"; {grid_honesty}"
    )


__all__ = [
    "DEFAULT_ROUTE_WEIGHTS",
    "MODEL_PRICES",
    "ROUTING_PREVIEW_DISCLOSURE",
    "SCORE_TIE_EPSILON",
    "InvalidCandidateError",
    "InvalidRouteWeightsError",
    "MissingPriceError",
    "RoutingCandidate",
    "RoutingDecision",
    "ScoredCandidate",
    "candidate_id",
    "normalize_route_weights",
    "parse_candidate",
    "parse_candidates",
    "preview_routing",
    "price_for_model",
    "score_candidates",
]

File: packages/core-ts/src/routing.ts
/**
 * Cross-provider LLM routing — the scored-candidate selection engine
 * (ROADMAP item 1).
 *
 * Given a set of caller-allowed `provider:model` candidates and the grid
 * intensity at the ALREADY-CHOSEN dispatch window, pick the best candidate on
 * a weighted blend of carbon, cost and latency. The window is selected once
 * by `selectWindow` BEFORE this runs — routing never shops for a per-candidate
 * window.
 *
 * ## Honesty constraints (non-negotiable, mirrored in the reasoning string)
 *
 *  1. **No silent model swaps.** Routing only ever picks among the candidates
 *     the caller explicitly allowed. There are no built-in model-equivalence
 *     tiers — a one-candidate list routes to exactly that candidate.
 *  2. **Grid honesty.** For hosted providers we do NOT know the data-centre
 *     grid; every hosted candidate is scored against the SAME caller's-grid
 *     intensity (the receipt's documented assumption). Routing differentiates
 *     carbon via per-model SSOT energy coefficients, not pretend per-provider
 *     grids. Ollama is genuinely local. The reasoning string says which.
 *  3. **Explainable pick.** The full scored candidate list + the normalized
 *     weights are returned (and recorded on the receipt); the pick is
 *     reproducible given the same rng seed.
 *
 * ## Scoring
 *
 * For each candidate `c` at the chosen window `W`:
 *  - `carbonG(c)  = estimateEnergyKwh(model, tokens) × intensity(W)` — the
 *    same math today's receipts use.
 *  - `costUsd(c)  = tokens × price(model)`, halved by `batchDiscount` when the
 *    task is batch-eligible AND the adapter supports batch. A candidate model
 *    missing from the price table is REJECTED loudly (never guessed).
 *  - `latencyClass(c)` — static tiers: local (ollama) = 0, hosted-sync = 0.5,
 *    hosted-batch = 1. No fake milliseconds.
 *
 * Each dimension is normalized across the candidates to `[0,1]` (min→0, max→1;
 * all-equal→0). `score = w_carbon·carbon + w_cost·cost + w_latency·latency`.
 * Lowest score wins; exact ties are broken with the injectable rng.
 */

import { estimateEnergyKwh, normalizeModelName } from "./energy.js";
import { MODEL_PRICES, TYPICAL_INPUT_TOKENS, TYPICAL_OUTPUT_TOKENS } from "./data/tables.generated.js";
import type { ProviderName } from "./types.js";

/** One row of the SSOT price table (`data/prices.json`). USD per million tokens. */
export interface ModelPrice {
  /** USD per million input (prompt) tokens (public list price). */
  inUsdPerMtok: number;
  /** USD per million output (completion) tokens (public list price). */
  outUsdPerMtok: number;
  /** Multiplier applied when the candidate runs via the vendor Batch API
   *  (0.5 for Anthropic / OpenAI). Absent ⇒ no batch discount available. */
  batchDiscount?: number;
  /** Month the figure was read, e.g. "2026-07". */
  asOf: string;
  /** Vendor pricing-page citation. */
  source: string;
}

/** The three routing dimensions and their relative importance. Non-negative. */
export interface RouteWeights {
  carbon: number;
  cost: number;
  latency: number;
}

/** Default weights when the caller does not supply `route_weights`. */
export const DEFAULT_ROUTE_WEIGHTS: RouteWeights = Object.freeze({
  carbon: 0.6,
  cost: 0.3,
  latency: 0.1,
});

/** A parsed `provider:model` candidate. */
export interface RoutingCandidate {
  provider: ProviderName;
  model: string;
}

/**
 * Static latency tier — deliberately coarse, no fake milliseconds:
 *   - `0`   local (Ollama)
 *   - `0.5` hosted synchronous call
 *   - `1`   hosted Batch API (up-to-24h SLA)
 */
export type LatencyClass = 0 | 0.5 | 1;

/** One fully-scored candidate, as recorded on the receipt's routing block. */
export interface ScoredCandidate {
  provider: ProviderName;
  model: string;
  /** Estimated grams CO2e for this candidate at the chosen window. */
  estCarbonG: number;
  /** Estimated USD list cost for this candidate (batch discount applied when it applies). */
  estCostUsd: number;
  latencyClass: LatencyClass;
  /** Weighted, normalized blend in [0,1]. Lower is better. */
  score: number;
}

/**
 * The routing decision — persisted on the task row at schedule time and
 * embedded in the signed receipt at completion. `chosen` / `fallbackFrom`
 * are compact `"provider:model"` ids.
 */
export interface RoutingDecision {
  weights: RouteWeights;
  considered: ScoredCandidate[];
  chosen: string;
  /** Set when dispatch fell back off the originally-chosen candidate. */
  fallbackFrom?: string;
  /** One honest human-readable line for an LLM caller / a receipt renderer. */
  reasoning: string;
}

/**
 * A NON-BINDING routing preview, returned by `recommend_window` and
 * `schedule_task` dry_run. Same scored-candidate math as the committing path,
 * but scored at the PREVIEWED window's intensity — the binding pick is made
 * at schedule time and may differ if the forecast shifts before commit. The
 * `preview` flag and the reasoning prefix disclose that plainly.
 */
export interface RoutingPreview extends RoutingDecision {
  preview: true;
}

/** Disclosure prefix on a preview's reasoning line. */
export const ROUTING_PREVIEW_DISCLOSURE =
  "PREVIEW — the binding pick is decided at schedule time and may differ if the forecast shifts before commit";

/** The providers a candidate string may name (keep in sync with {@link ProviderName}). */
const KNOWN_PROVIDERS: ReadonlySet<string> = new Set([
  "anthropic",
  "openai",
  "gemini",
  "ollama",
]);

/** Providers whose adapters can route through a synchronous vendor Batch API. */
const BATCH_CAPABLE_PROVIDERS: ReadonlySet<ProviderName> = new Set([
  "anthropic",
  "openai",
]);

/** Thrown when a `provider:model` candidate string is malformed. */
export class InvalidCandidateError extends Error {
  constructor(public readonly received: string, detail: string) {
    super(`Invalid routing candidate ${JSON.stringify(received)}: ${detail}`);
    this.name = "InvalidCandidateError";
  }
}

/**
 * Thrown when one or more candidate models are absent from the SSOT price
 * table. Loud by design — routing never guesses a price. Lists every missing
 * id so the caller can fix the whole candidate set at once.
 */
export class MissingPriceError extends Error {
  constructor(public readonly missing: string[]) {
    super(
      `No price in data/prices.json for routing candidate model(s): ${missing.join(", ")}. ` +
        `Add each to prices.json (with asOf + source) or drop it from the candidate list — routing never guesses a price.`,
    );
    this.name = "MissingPriceError";
  }
}

/** Thrown when the caller-supplied weights cannot be normalized. */
export class InvalidRouteWeightsError extends Error {
  constructor(detail: string) {
    super(`Invalid route_weights: ${detail}`);
    this.name = "InvalidRouteWeightsError";
  }
}

/** Parse a single `"provider:model"` string. */
export function parseCandidate(spec: string): RoutingCandidate {
  if (typeof spec !== "string") {
    throw new InvalidCandidateError(String(spec), "expected a string");
  }
  const trimmed = spec.trim();
  const idx = trimmed.indexOf(":");
  if (idx <= 0) {
    throw new InvalidCandidateError(spec, 'expected "provider:model"');
  }
  const provider = trimmed.slice(0, idx).trim().toLowerCase();
  const model = trimmed.slice(idx + 1).trim();
  if (!KNOWN_PROVIDERS.has(provider)) {
    throw new InvalidCandidateError(
      spec,
      `unknown provider ${JSON.stringify(provider)} (expected one of anthropic, openai, gemini, ollama)`,
    );
  }
  if (model.length === 0) {
    throw new InvalidCandidateError(spec, "model part is empty");
  }
  return { provider: provider as ProviderName, model };
}

/** Parse a list of `"provider:model"` strings. */
export function parseCandidates(specs: readonly string[]): RoutingCandidate[] {
  return specs.map(parseCandidate);
}

/** Compact `"provider:model"` id for a candidate. */
export function candidateId(c: { provider: string; model: string }): string {
  return `${c.provider}:${c.model}`;
}

/**
 * Normalize caller weights to a non-negative vector summing to 1. Absent ⇒
 * {@link DEFAULT_ROUTE_WEIGHTS}. Missing keys default to 0; any negative value
 * or an all-zero vector is rejected.
 */
export function normalizeRouteWeights(w?: Partial<RouteWeights>): RouteWeights {
  if (w === undefined || w === null) {
    return { ...DEFAULT_ROUTE_WEIGHTS };
  }
  const carbon = w.carbon ?? 0;
  const cost = w.cost ?? 0;
  const latency = w.latency ?? 0;
  for (const [k, v] of [["carbon", carbon], ["cost", cost], ["latency", latency]] as const) {
    if (typeof v !== "number" || Number.isNaN(v)) {
      throw new InvalidRouteWeightsError(`${k} must be a number, got ${JSON.stringify(v)}`);
    }
    if (v < 0) {
      throw new InvalidRouteWeightsError(`${k} must be non-negative, got ${v}`);
    }
  }
  const sum = carbon + cost + latency;
  if (sum <= 0) {
    throw new InvalidRouteWeightsError("at least one weight must be positive");
  }
  return { carbon: carbon / sum, cost: cost / sum, latency: latency / sum };
}

/** Look up the SSOT price for a model id, trying exact then normalized. */
export function priceForModel(model: string): ModelPrice | undefined {
  const exact = MODEL_PRICES[model.trim().toLowerCase()];
  if (exact) return exact;
  return MODEL_PRICES[normalizeModelName(model)];
}

/** Whether a candidate would run via the vendor Batch API given task/adapter state. */
function willRunBatch(
  c: RoutingCandidate,
  price: ModelPrice,
  batchEligible: boolean,
  batchCapable: (provider: ProviderName) => boolean,
): boolean {
  return (
    batchEligible &&
    BATCH_CAPABLE_PROVIDERS.has(c.provider) &&
    price.batchDiscount !== undefined &&
    batchCapable(c.provider)
  );
}

function latencyClassFor(c: RoutingCandidate, runsBatch: boolean): LatencyClass {
  if (c.provider === "ollama") return 0;
  return runsBatch ? 1 : 0.5;
}

/** USD list cost for a candidate at the given token shape. */
function costUsdFor(
  price: ModelPrice,
  inputTokens: number,
  outputTokens: number,
  runsBatch: boolean,
): number {
  const base =
    (inputTokens / 1_000_000) * price.inUsdPerMtok +
    (outputTokens / 1_000_000) * price.outUsdPerMtok;
  const mult = runsBatch && price.batchDiscount !== undefined ? price.batchDiscount : 1;
  return base * mult;
}

export interface ScoreCandidatesOptions {
  /** Parsed candidate list (>= 1). >= 2 is where routing actually differentiates. */
  candidates: readonly RoutingCandidate[];
  /** Grid intensity (gCO2e/kWh) at the already-chosen dispatch window. */
  intensityGCo2PerKwh: number;
  /** Caller weights; normalized internally. Absent ⇒ DEFAULT_ROUTE_WEIGHTS. */
  weights?: Partial<RouteWeights>;
  /** True when (deadline − now) > 24h, i.e. the Batch API SLA still fits. */
  batchEligible?: boolean;
  /**
   * Whether a configured adapter for `provider` supports batch this run.
   * Defaults to "yes for anthropic/openai" — the scheduler passes a stricter
   * predicate keyed on the actually-configured adapters.
   */
  batchCapable?: (provider: ProviderName) => boolean;
  /** Input tokens for the estimate. Defaults to the typical-task shape. */
  inputTokens?: number;
  /** Output tokens for the estimate. Defaults to the typical-task shape. */
  outputTokens?: number;
  /** Injectable rng in [0,1) for the tie-break. Defaults to Math.random. */
  rng?: () => number;
}

/** Ties within this score epsilon are broken with the rng (reproducible given a seed). */
export const SCORE_TIE_EPSILON = 1e-9;

/**
 * Score a candidate set and pick the lowest-scoring one. Throws
 * {@link MissingPriceError} if any candidate model is absent from the price
 * table (loud, never guessed).
 */
export function scoreCandidates(opts: ScoreCandidatesOptions): RoutingDecision {
  const {
    candidates,
    intensityGCo2PerKwh,
    batchEligible = false,
    batchCapable = (p) => BATCH_CAPABLE_PROVIDERS.has(p),
    inputTokens = TYPICAL_INPUT_TOKENS,
    outputTokens = TYPICAL_OUTPUT_TOKENS,
    rng = Math.random,
  } = opts;
  if (candidates.length === 0) {
    throw new Error("scoreCandidates: candidates must be non-empty");
  }
  const weights = normalizeRouteWeights(opts.weights);

  // Resolve every price FIRST so a missing one rejects the whole set loudly.
  const missing: string[] = [];
  const prices = candidates.map((c) => {
    const p = priceForModel(c.model);
    if (!p) missing.push(candidateId(c));
    return p;
  });
  if (missing.length > 0) {
    throw new MissingPriceError(missing);
  }

  // Raw per-dimension figures.
  const raw = candidates.map((c, i) => {
    const price = prices[i]!;
    const runsBatch = willRunBatch(c, price, batchEligible, batchCapable);
    const carbonG =
      estimateEnergyKwh({ model: c.model, inputTokens, outputTokens }) * intensityGCo2PerKwh;
    const costUsd = costUsdFor(price, inputTokens, outputTokens, runsBatch);
    const latencyClass = latencyClassFor(c, runsBatch);
    return { c, carbonG, costUsd, latencyClass };
  });

  const norm = (values: number[]): number[] => {
    const min = Math.min(...values);
    const max = Math.max(...values);
    if (max - min <= 0) return values.map(() => 0);
    return values.map((v) => (v - min) / (max - min));
  };
  const nCarbon = norm(raw.map((r) => r.carbonG));
  const nCost = norm(raw.map((r) => r.costUsd));
  const nLatency = norm(raw.map((r) => r.latencyClass));

  const considered: ScoredCandidate[] = raw.map((r, i) => ({
    provider: r.c.provider,
    model: r.c.model,
    estCarbonG: round4(r.carbonG),
    estCostUsd: round6(r.costUsd),
    latencyClass: r.latencyClass,
    score: round6(
      weights.carbon * nCarbon[i]! + weights.cost * nCost[i]! + weights.latency * nLatency[i]!,
    ),
  }));

  // Lowest score wins; exact ties (within epsilon) broken by rng. Reproducible
  // given a seeded rng — same tolerance spirit as selectWindow.
  const minScore = Math.min(...considered.map((c) => c.score));
  const tied = considered.filter((c) => c.score - minScore <= SCORE_TIE_EPSILON);
  const chosen =
    tied.length > 1 ? tied[Math.floor(rng() * tied.length)]! : tied[0]!;

  return {
    weights,
    considered,
    chosen: candidateId(chosen),
    reasoning: buildRoutingReasoning(considered, chosen, weights),
  };
}

/**
 * Non-binding routing preview for the planning surfaces (`recommend_window`,
 * `schedule_task` dry_run). Runs the SAME scoring as {@link scoreCandidates}
 * at the previewed window's intensity, then marks the result as a preview and
 * prefixes the reasoning with {@link ROUTING_PREVIEW_DISCLOSURE}. Returns
 * `undefined` when fewer than two candidates were supplied (routing is a no-op
 * — no block is shown, so the params are never inert: they take effect exactly
 * when >= 2 candidates are present). Throws {@link MissingPriceError} /
 * {@link InvalidCandidateError} loudly, same as the committing path.
 */
export function previewRouting(
  candidates: readonly string[] | undefined,
  opts: Omit<ScoreCandidatesOptions, "candidates">,
): RoutingPreview | undefined {
  if (!Array.isArray(candidates) || candidates.length < 2) return undefined;
  const decision = scoreCandidates({ ...opts, candidates: parseCandidates(candidates) });
  return {
    ...decision,
    preview: true,
    reasoning: `${ROUTING_PREVIEW_DISCLOSURE}: ${decision.reasoning}`,
  };
}

// --------------------------------------------------------------------------- //
// Reasoning

function buildRoutingReasoning(
  considered: ScoredCandidate[],
  chosen: ScoredCandidate,
  weights: RouteWeights,
): string {
  const others = considered.filter((c) => c !== chosen);
  const vs = others.length
    ? ` (score ${fmtScore(chosen.score)} vs ${others.map((o) => fmtScore(o.score)).join("/")})`
    : "";

  // Which dimension the chosen candidate leads on (honest: only claim "lowest"
  // when it truly is the min, and only mention dimensions with weight).
  const clauses: string[] = [];
  if (weights.carbon > 0 && isMinBy(considered, chosen, (c) => c.estCarbonG)) {
    clauses.push("lowest carbon at window");
  }
  if (weights.cost > 0 && isMinBy(considered, chosen, (c) => c.estCostUsd)) {
    clauses.push(chosen.estCostUsd === 0 ? "no per-token cost (local)" : "lowest cost");
  }
  if (weights.latency > 0 && isMinBy(considered, chosen, (c) => c.latencyClass)) {
    clauses.push(chosen.latencyClass === 0 ? "local (no network latency tier)" : "fastest latency tier");
  }
  if (clauses.length === 0) {
    clauses.push("best weighted blend of carbon/cost/latency");
  }

  const gridHonesty =
    chosen.provider === "ollama"
      ? "runs locally on your own grid"
      : "hosted-grid assumption applies (all hosted candidates scored at the same caller's-grid intensity)";

  return `routed to ${candidateId(chosen)}${vs}: ${clauses.join("; ")}; ${gridHonesty}`;
}

function isMinBy(
  all: ScoredCandidate[],
  target: ScoredCandidate,
  key: (c: ScoredCandidate) => number,
): boolean {
  const min = Math.min(...all.map(key));
  return key(target) <= min + SCORE_TIE_EPSILON;
}

/** Format a [0,1] score like ".21" (leading zero stripped), matching the design's examples. */
function fmtScore(x: number): string {
  const s = x.toFixed(2);
  return s.startsWith("0.") ? s.slice(1) : s;
}

function round4(v: number): number {
  return Math.round(v * 10000) / 10000;
}
function round6(v: number): number {
  return Math.round(v * 1000000) / 1000000;
}
